### ISSUE TITLE
Problems with DOGMA data preprocessing prevent good reproduction

### ISSUE THREAD

**Issue by pangdundun** [labomics/midas#19]
Hello, I encountered a problem in reproducing MIDAS on the dogma_diagonal_a combination. The UMAP cluster figure shows that the problem seems to be with the ATAC data, the details are as follows.

Target: dogma_diagonal_a combination, dogma_paired_a

**Command of Data processing:**

> Rscript preprocess/atac+rna+adt_dogma_lll_ctrl.R
Rscript preprocess/atac+rna+adt_dogma_lll_stim.R
Rscript preprocess/atac+rna+adt_dogma_dig_ctrl.R
Rscript preprocess/atac+rna+adt_dogma_dig_stim.R
Rscript preprocess/annotate_seurat.R
Rscript preprocess/combine_subsets.R --task dogma_diagonal_a
python preprocess/split_mat.py --task dogma_diagonal_a
Rscript preprocess/combine_subsets.R --task dogma_paired_a
python preprocess/split_mat.py --task dogma_paired_a

**Command of training MIDAS:**

> data = [GetDataInfo(f'./data/processed/dogma_diagonal_a')]
model.train(n_epoch=500, save_path=f'./result/dogma_diagonal_a/train/')
model.predict(save_dir=f'./result/dogma_diagonal_a/predict/', mod_latent=True)

**MIDAS results in the dogma_diagonal_a combination (500 epoch):**

Code obtained from [Tutorial-2](https://scmidas.readthedocs.io/en/latest/notebooks/tutorial_2.html#).

![e500](https://github.com/user-attachments/assets/77858d66-0ee5-4254-b0b3-f583a1214659)
![e500](https://github.com/user-attachments/assets/973b7815-8caa-475c-a223-501acc06a48f)
![e500](https://github.com/user-attachments/assets/b28e4129-b4e1-4edf-9820-56f60f8da169)

**GLUE results in the dogma_diagonal_a combination (500 epoch):**

![e500](https://github.com/user-attachments/assets/56bed111-bbae-41f7-9e90-f5cd8142f796)
![e500](https://github.com/user-attachments/assets/1dede682-9b85-4f87-ae75-191d0164f398)
![e500](https://github.com/user-attachments/assets/4f61a314-fab1-44a6-b68a-905ee1b68b5b)

**MIDAS results in the dogma_paired_a combination (125 epoch):**

![e125](https://github.com/user-attachments/assets/6b071300-29c8-433c-b233-edd91b1cab2a)
![e125](https://github.com/user-attachments/assets/bb32797d-2513-494f-a50c-9542ac0cf999)

**Compared with the appendix of the paper:**

Compared with the supplementary figures 16 and 17 of the paper, found that the overall_score is very low, clustering is not good, especially for ATAC, and the clustering under cell type coloring show a messy distribution.

![Supplementary figures 16.b](https://github.com/user-attachments/assets/a10f09e8-252d-4e7f-90ce-7944078faf8e)
![Supplementary figures 17.c](https://github.com/user-attachments/assets/082368fc-153b-4615-8969-714d77b232d2)

**MIDAS results in the dogma_demo combination (50 epoch):**

Use [the preprocessed data](https://drive.google.com/drive/folders/1j3hDl2p4v3s0r5a-ktfHqbIhHbIPpUbS) provided by the author for training, the results are good.

![e50](https://github.com/user-attachments/assets/7611d975-f692-44a7-bf3c-8cd2f1721c67)
![e50](https://github.com/user-attachments/assets/afc33ee7-01a1-46b1-a989-6afdad2f2356)

**My question:**

Because in the clustering of modality embedding, the cell type coloring clusters of ATAC in dogma_demo have clear boundaries, but the cell type coloring clusters of ATAC in the ones I processed, such as dogma_diagonal_a and dogma_paired_a, are very confusing. So I think the problem lies in the preprocessing of ATAC data, how do you think?

And I want to know how can I correct this problem so that MIDAS and GLUE cound reach the expected 0.5 and 0.4 for the dogma_diagonal_a combination.

Thanks! And look forward your reply.

**Comment by JiahaoChow** [comment]
Hello,

Thank you for your interest in reproducing our results. Here are a few suggestions to help you:

1. Please use the `reproducibility` branch to reproduce our results, as we have not yet fully tested all experiments in our paper with the latest version of the code.  

2. **Important Note**: The results may vary depending on the number of training epochs. In our paper, most results were based on 2000 training epochs. However, in our latest tutorials, we used a setting of 500 epochs to demonstrate a quick run. We apologize for not clarifying this earlier.

3. For the reproducibility of GLUE, please ensure you use the guidance graph from [[this link]](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_43/gencode.v43.annotation.gtf.gz). To adapt the GLUE tutorials, you can directly replace the inputs for ATAC and RNA with the generated `atac.csv` and `rna.csv` files, which should already be available after running our preprocessing scripts (or maybe you could use the common features for RNA data).

If you have any further questions or encounter issues, feel free to reach out.

Best regards

**Comment by pangdundun** [comment]
> Hello,
> 
> Thank you for your interest in reproducing our results. Here are a few suggestions to help you:
> 
> 1. Please use the `reproducibility` branch to reproduce our results, as we have not yet fully tested all experiments in our paper with the latest version of the code.
> 2. **Important Note**: The results may vary depending on the number of training epochs. In our paper, most results were based on 2000 training epochs. However, in our latest tutorials, we used a setting of 500 epochs to demonstrate a quick run. We apologize for not clarifying this earlier.
> 3. For the reproducibility of GLUE, please ensure you use the guidance graph from [[this link]](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_43/gencode.v43.annotation.gtf.gz). To adapt the GLUE tutorials, you can directly replace the inputs for ATAC and RNA with the generated `atac.csv` and `rna.csv` files, which should already be available after running our preprocessing scripts (or maybe you could use the common features for RNA data).
> 
> If you have any further questions or encounter issues, feel free to reach out.
> 
> Best regards

Hi,

You think the reason why MIDAS and GLUE have Poor reproduction results is that I used inappropriate model codes, set incomplete training epochs, or used lower version of genome annotation. However, I think the problem lies not on the model side, but on the data side:

1. Just from the GLUE point of view, the paper gives a score of 0.409. But, Whether I use v35 or v43 that you provided, the score fluctuates around 0.29, which still is a big gap from 0.409. Comparing the joint embedding UMAP graph with paper supplementary figure, gap is obvious...

2. Similar to GLUE, I was unable to achieve the paper’s 0.500 score on DOGMA diagonal_a with MIDAS, again by a large margin, even though I only ran 1000 epochs.

3. Replaced the DOGMA cell type labels obtained by my own preprocessing with [the labels in dogma-demo](https://drive.google.com/drive/folders/1j3hDl2p4v3s0r5a-ktfHqbIhHbIPpUbS) provided by the author. Although the scores fluctuated, I still could not get the expected scores of dogma_paired_a and dogma_diagonal_a.

4. Used my own data propressed from running combine_subsets.R, the results was poor. However, the results was good when used dogma_demo your provided.

I think the above is a good indication that the problem comes from the data rather than the model or cell type label. I'm new to data processing, and don't know where the problem comes from. So I would like to get your help as follows:

1. Could you get possible insights into the problem from the two pictures below? That are the joint embedding UMAP and single modality embedding UMAP of MIDAS for DOGMA diagonal_a combination at the 275th, 500th, 1000th epochs, and Supplementary Figure 17.c (MIDAS) respectively.

2. I understand that providing pre-processed data files is out of your job and is time-consuming and laborious, but I still want to know if you are allowed to provide the data files processed by combine_subsets.R, such as dogma_demo_processed.zip.

Best regards

![20241222-214536](https://github.com/user-attachments/assets/18ac7d4e-ca54-46a4-bc66-4a2da04e2c7b)
![20241222-202410](https://github.com/user-attachments/assets/2903f50f-b84f-415c-bffd-776e0119b810)

**Comment by JiahaoChow** [comment]
Hello,

We noticed that you are using the `dogma_diagonal_a` task. Kindly try using the `dogma_diagonal_b` task instead, as this is the task we actually used in the paper. It is also a diagonal task involving RNA and ATAC data.

If you are still unable to reproduce the results, we would be happy to provide you with our processed data or any additional resources that might help.

Best regards,

**Comment by JiahaoChow** [comment]
For bimodal experiments, please follow these steps:

1. Use the model configuration `l_2`.
2. For scmidas (versions < 0.1), set `disc_train=5` in the `init_model()` function.
For the reproducibility branch, you can set the experiment to `k_5`, which automatically sets `k` to 5. For example:
`python run.py --model l_2 --experiment k_5 --task dogma_diagonal_b`

**Comment by pangdundun** [comment]
> Hello,
> 
> We noticed that you are using the `dogma_diagonal_a` task. Kindly try using the `dogma_diagonal_b` task instead, as this is the task we actually used in the paper. It is also a diagonal task involving RNA and ATAC data.
> 
> If you are still unable to reproduce the results, we would be happy to provide you with our processed data or any additional resources that might help.
> 
> Best regards

Hi,

Sadly, I still cannot reproduce the valid results.

Followed your instructions, adjusted the RNA encoder core from [1024, 128] to [1024], and the discriminator iterations from 3 to 5. However, no better results were obtained for the dogma_paired_a and dogma_diagonal_b.

The changes in their overall scores in epochs 1800 and 2000 are shown in the following two figures. The former is 0.626, cannot reach 0.784. The latter is 0.418, cannot reach 0.503. And the overall score cannot increase with the increase of epoch.

Considering that GLUE is 0.290 and cannot reach 0.401, **it is hoped that the preprocessed data can be obtained**. Specifically, **the five bimodal mosaic combinations** used in Supplementary Figure 16.c.

Thanks for your continued attention.

Best regards

![20241226-185412](https://github.com/user-attachments/assets/848e612b-bc40-4c5c-a8c8-2a70dd25669b)
![20241226-185408](https://github.com/user-attachments/assets/424af37b-1ac1-479e-9e1c-4b114750d985)

**Comment by zhen-he** [comment]
Hi, please check the [attached files](https://drive.google.com/file/d/1KueQj7OC6zdyuUvm_wvM0XLj7MiEzLtp/view?usp=sharing) to reproduce the results.

- `data/processed` contains the processed data. You need to run `split_mat.py` to generated csv files for training.
- `data/raw` contains the Seurat labels for scIB evaluation.
- `result/dogma_*/k_5/l_2/train` contains the trained models.
- `result/dogma_*/k_5/l_2/predict` contains the predicted joint embeddings [**_c_**, **_u_**] for scIB evaluation.
